File: go/callwire/registry_test.go
package callwire

import (
	"net"
	"os"
	"testing"
	"time"
)

func TestServiceDiscovery(t *testing.T) {
	// Disable auto-serve for this test.
	os.Setenv("CALLWIRE_AUTO", "0")
	defer os.Unsetenv("CALLWIRE_AUTO")

	// Start registry server on a random port.
	regListener, err := ServeRegistry("127.0.0.1:0")
	if err != nil {
		t.Fatalf("failed to start registry server: %v", err)
	}
	defer regListener.Close()
	regAddr := regListener.Addr().String()

	// Register some functions on a worker server.
	Export("say_hello", func(name string) (string, error) {
		return "Hello, " + name, nil
	})

	// Start worker server on a dynamic port.
	workerListener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("failed to listen worker: %v", err)
	}
	workerAddr := workerListener.Addr().String()
	workerListener.Close()
	
	go func() {
		_ = Serve(workerAddr)
	}()
	time.Sleep(50 * time.Millisecond) // let worker start

	// Register worker with registry.
	err = RegisterWith(regAddr, "hello-service", workerAddr)
	if err != nil {
		t.Fatalf("failed to register hello-service: %v", err)
	}

	// Create discover pool.
	pool, err := NewDiscoverPool(regAddr, "hello-service")
	if err != nil {
		t.Fatalf("failed to create discover pool: %v", err)
	}
	defer pool.Close()

	sayHello := DiscoverRef[string](pool, "say_hello")
	reply, err := sayHello("World")
	if err != nil {
		t.Fatalf("RPC call failed: %v", err)
	}
	if reply != "Hello, World" {
		t.Errorf("expected 'Hello, World', got %q", reply)
	}
}

File: rust/tests/registry_test.rs
use callwire::{Client, DiscoverPool, serve_registry, serve_on, register_unary};

#[tokio::test]
async fn test_rust_registry_discovery() {
    unsafe { std::env::set_var("CALLWIRE_AUTO", "0"); }

    // Start registry
    let reg_handle = serve_registry("127.0.0.1:29090").await.expect("failed to start registry");
    tokio::time::sleep(std::time::Duration::from_millis(50)).await;

    // Export function on worker
    register_unary("say_hello", |(name,): (String,)| -> Result<String, String> {
        Ok(format!("Hello {}", name))
    });

    // Start worker
    let worker_handle = serve_on("127.0.0.1:29091").await.expect("failed to start worker");
    tokio::time::sleep(std::time::Duration::from_millis(50)).await;

    // Register worker
    let reg_client = Client::connect("127.0.0.1:29090").await.expect("failed to connect to registry");
    let _: () = reg_client.import("callwire.register", &("test-service".to_string(), "127.0.0.1:29091".to_string())).await.unwrap();

    // Discover and call
    let pool = DiscoverPool::new("127.0.0.1:29090", "test-service").await.expect("failed to create pool");
    let client = pool.get().expect("failed to get client");
    
    let res: String = client.import("say_hello", &("World".to_string(),)).await.expect("call failed");
    assert_eq!(res, "Hello World");

    reg_handle.close();
    worker_handle.close();
}
